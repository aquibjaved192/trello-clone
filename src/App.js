import React, { useState, useCallback } from 'react';
import Card from './components/card';
import style from './App.module.scss';

function App() {
  const [cards, setCards] = useState([{id: Date.now() ,name: "Enter Card Name", tasks: []}]);

  const handleSubmit = () => {
    const card = { 
      id: Date.now(),
      name: "Enter Card Name",
      tasks: [],
    }
    setCards((cards) => [...cards, card]);
  }

  const editCard = useCallback((data, id, operation) => {
    const newCards = [...cards]
    newCards.forEach(item => {
      if(item.id === id) {
        if(operation === 'name') {
          item.name = data
        } else if(operation === 'addTask') {
          item.tasks.push(data);
        }
      }
    })
    setCards(newCards);
  },[cards]);

  const editTask = useCallback((data, cardId, taskIndex) => {
    const newCards = [...cards]
    newCards.forEach(item => {
      if(item.id === cardId) {
        if(data === null) {
          item.tasks.splice(taskIndex, 1);
        } else {
          item.tasks.splice(taskIndex, 1, data);
        }
      }
    })
    setCards(newCards);
  },[cards]);

  const deleteCard = useCallback((id) => {
    const oldCards = [...cards];
    const newCards = oldCards.filter(item => item.id !== id);
    setCards(newCards);
  },[cards]);

  const handleDrop = useCallback((e, data) => {
    let task = JSON.parse(e.dataTransfer.getData("task"));
    if(task.parentId !== data.id) {
      let newCards = [...cards];
    
      newCards.forEach((card) => {   
        if (card.id === task.parentId) {    
          let newTasks = Object.assign(card.tasks);           
          newTasks.splice(task.Index, 1);  
          card.tasks = newTasks;              
        } 
        
        if(card.id === data.id) {
          const newTasks = Object.assign(card.tasks);
          const pushTask = { parentId : data.id, id: task.id, taskName: task.taskName }
          newTasks.push(pushTask);
          card.tasks = newTasks;
        }
      });
      
      setCards(newCards)
    }  
  },[cards])

  return (
    <div className={style.App}>
      {cards.map((item) => 
        <Card
          key={item?.id}
          data={item}
          editCard={editCard}
          editTask={editTask}
          deleteCard={deleteCard}
          handleDrop={handleDrop}
          length={cards.length}
        />  
      )}
      <div>
        <button data-testid="addCard" onClick={handleSubmit} className={style.addCard}>+</button>
      </div>
    </div>
  );
}

export default App;
