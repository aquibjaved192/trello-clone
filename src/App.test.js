import React from 'react'
import { render, cleanup, fireEvent, screen } from '@testing-library/react';
import App from './App'

afterEach(cleanup)
 
it('should take a snapshot', () => {
  const { asFragment } = render(<App />)
  expect(asFragment(<App />)).toMatchSnapshot()
})

it('should render a add card button', () => {
  const { getByTestId } = render(<App />); 
  expect(getByTestId('addCard')).toBeInTheDocument();
});

it('should add a card on click add card button', () => {
  const { getByTestId } = render(<App />); 
  
  fireEvent.click(getByTestId('addCard'));

  const deleteCards = screen.getAllByTestId('deleteCard');

  expect(deleteCards).toHaveLength(2);
});