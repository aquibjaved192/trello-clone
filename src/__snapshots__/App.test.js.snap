// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="card"
      data-testid="card-container"
    >
      <div
        class="label"
      >
        <label
          class="editLabel"
          data-testid="label"
        >
          Enter Card Name
        </label>
      </div>
      <form>
        <input
          class="inputTask"
          data-testid="taskInput"
          placeholder="Add Task"
          required=""
          type="text"
          value=""
        />
        <button
          class="addTask"
          data-testid="addTask"
          type="submit"
        >
          +
        </button>
      </form>
      <div>
        <button
          class="delete"
          data-testid="deleteCard"
          disabled=""
        >
          Delete
        </button>
      </div>
    </div>
    <div>
      <button
        class="addCard"
        data-testid="addCard"
      >
        +
      </button>
    </div>
  </div>
</DocumentFragment>
`;
